server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java

```diff
@@ -0,0 +1,193 @@
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.cluster.ClusterState.INFERRED_TRANSPORT_VERSION;
+
+/**
+ * This fixes up the transport version from pre-8.8.0 cluster state that was inferred as the minimum possible,
+ * due to the master node not understanding cluster state with transport versions added in 8.8.0.
+ * Any nodes with the inferred placeholder cluster state is then refreshed with their actual transport version
+ */
+public class TransportVersionsFixupListener implements ClusterStateListener {
+
+    private static final Logger Log = LogManager.getLogger(TransportVersionsFixupListener.class);
+
+    private static final TimeValue RETRY_TIME = TimeValue.timeValueSeconds(30);
+    private static final int MAX_RETRIES = 120; // try for 1 hour
+
+    private final MasterServiceTaskQueue<NodeTransportVersionTask> taskQueue;
+    private final ClusterAdminClient client;
+    private final Scheduler scheduler;
+    private final Set<String> pendingNodes = Collections.synchronizedSet(new HashSet<>());
+
+    public TransportVersionsFixupListener(ClusterService service, ClusterAdminClient client, Scheduler scheduler) {
+        this(service.createTaskQueue("fixup-transport-versions", Priority.LOW, new TransportVersionUpdater()), client, scheduler);
+    }
+
+    TransportVersionsFixupListener(
+        MasterServiceTaskQueue<NodeTransportVersionTask> taskQueue,
+        ClusterAdminClient client,
+        Scheduler scheduler
+    ) {
+        this.taskQueue = taskQueue;
+        this.client = client;
+        this.scheduler = scheduler;
+    }
+
+    class NodeTransportVersionTask implements ClusterStateTaskListener {
+        private final Map<String, TransportVersion> results;
+        private final int retryNum;
+
+        NodeTransportVersionTask(Map<String, TransportVersion> results, int retryNum) {
+            this.results = results;
+            this.retryNum = retryNum;
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            Log.error("Could not apply transport version for nodes {} to cluster state", results.keySet(), e);
+            scheduleRetry(results.keySet(), retryNum);
+        }
+
+        public Map<String, TransportVersion> results() {
+            return results;
+        }
+    }
+
+    private static class TransportVersionUpdater implements ClusterStateTaskExecutor<NodeTransportVersionTask> {
+        @Override
+        public ClusterState execute(BatchExecutionContext<NodeTransportVersionTask> context) throws Exception {
+            ClusterState.Builder builder = ClusterState.builder(context.initialState());
+            for (var c : context.taskContexts()) {
+                for (var e : c.getTask().results().entrySet()) {
+                    // this node's transport version might have been updated already/node has gone away
+                    if (Objects.equals(builder.transportVersions().get(e.getKey()), INFERRED_TRANSPORT_VERSION)) {
```

Comment:
Are we certain the current value is not null? If so, can we assert that? If not, can we make sure we update it whenever that's not a no-op?

Reply:
Not sure what you mean? That's why I'm using `Objects.equals`, so the `Map.get` can return null and it'll just ignore it.

Reply:
Sure but we don't want to ignore a `null`, we want to fix it up with the correct version.

Reply:
Ah, you mean differentiating between the key not existing at all and it mapping to null? There's no situation where a `null` is a value in the transportversions map. So it'll either not be there at all (`get` will return null) -> we ignore it, or it'll be the inferred version -> we update it, or it's something else -> we leave it alone

Reply:
No, I meant a `null` arising from a missing key. Are you sure we want to ignore that case?

Reply:
It will be missing if the cluster state has already been updated to remove a node between the update which triggered the node-info fetch, and the task running to update the state. So we just ignore it in that case. Are there other situations this could be missing & the node is still part of the cluster?